src/mlpack/methods/hmm/hmm.hpp

```diff
@@ -290,6 +290,73 @@ class HMM
+   * @param prevForwardLogProb Vector in which forward probabilities for time
+   * t-1 will be saved. Passing prevForwardProb as an empty vector indicates the
+   * start of sequence or time t=0
+   * @param forwardLogProb Vector in which forward probabilities for time t
+   * will be saved.
+   * @return Log scale factor of the given sequence of emission at time t.
+   */
+  double LogScaleEmissionProb(const arma::vec& emissionLogProb,
+                       arma::vec& prevForwardLogProb,
+                       arma::vec& forwardLogProb) const;
+  /**
+   * Compute the log-likelihood of the given emission probability up to time t
+   *
+   * @param emissionLogProb emission probability at time t.
+   * @param logLikelihood Log-likelihood of the given sequence of emission
+   * probability up to time t-1
+   * @param prevForwardLogProb Vector in which forward probabilities for time
+   * t-1 will be saved. Passing prevForwardProb as an empty vector indicates the
+   * start of sequence or time t=0
+   * @param forwardLogProb Vector in which forward probabilities for time t
+   * will be saved.
+   * @return Log-likelihood of the given sequence of emission up to time t.
+   */
+  double LogLikelihoodEmissionProb(const arma::vec& emissionLogProb,
+                       double &logLikelihood,
+                       arma::vec& prevForwardLogProb,
+                       arma::vec& forwardLogProb) const;
+  /**
+   * Compute the log of the scaling factor of the given data at time t.
+   * To calculate the log-likelihood for the whole sequence, accumulate log
+   * scale over the entire sequence
+   *
+   * @param data observation at time t.
+   * @param prevForwardLogProb Vector in which forward probabilities for time
+   * t-1 will be saved. Passing prevForwardProb as an empty vector indicates the
+   * start of sequence or time t=0
+   * @param forwardLogProb Vector in which forward probabilities for time t
+   * will be saved.
+   * @return Log scale factor of the given sequence of data up at time t.
+   */
+  double LogScale(const arma::vec &data,
+                       arma::vec& prevForwardLogProb,
+                       arma::vec& forwardLogProb) const;
+  /**
+   * Compute the log-likelihood of the given data up to time t
+   *
+   * @param data observation at time t.
+   * @param logLikelihood Log-likelihood of the given sequence of data
+   * up to time t-1
+   * @param prevForwardLogProb Vector in which forward probabilities for time
+   * t-1 will be saved. Passing prevForwardProb as an empty vector indicates the
+   * start of sequence or time t=0
+   * @param forwardLogProb Vector in which forward probabilities for time t
+   * will be saved.
+   * @return Log-likelihood of the given sequence of data up to time t.
+   */
+  double LogLikelihood(const arma::vec &data,
+                       double &logLikelihood,
+                       arma::vec& prevForwardLogProb,
+                       arma::vec& forwardLogProb) const;
```

Comment:
Personally I think we should modify `prevForwardLogProb` in-place, giving only three parameters as input.  If the user wants to preserve `forwardLogProb` from one time step, they can copy it.  Let me know what you think of this idea.

Reply:
good suggestion. I will modify the code in this way.